File: WMS.DALCore/Producto/clsLnProducto_estado.cs
using Microsoft.Data.SqlClient;
using System.Data;
using System.Diagnostics;
using WMS.EntityCore.Producto;
using Microsoft.Extensions.Configuration;
public class clsLnProducto_estado
{
    private static readonly clsInsert Ins = new clsInsert();
    private static readonly clsUpdate Upd = new clsUpdate();

    public static void Cargar(ref clsBeProducto_estado oBe, DataRow dr)
    {
        oBe.IdEstado = dr.Field<int?>("IdEstado") ?? 0;
        oBe.IdPropietario = dr.Field<int?>("IdPropietario") ?? 0;
        oBe.Nombre = dr.Field<string>("nombre") ?? "";
        oBe.IdUbicacionDefecto = dr.Field<int?>("IdUbicacionDefecto") ?? 0;
        oBe.Utilizable = dr.Field<bool?>("utilizable") ?? false;
        oBe.Activo = dr.Field<bool?>("activo") ?? false;
        oBe.User_agr = dr.Field<string>("user_agr") ?? "";
        oBe.Fec_agr = dr.Field<DateTime?>("fec_agr") ?? DateTime.Now;
        oBe.User_mod = dr.Field<string>("user_mod") ?? "";
        oBe.Fec_mod = dr.Field<DateTime?>("fec_mod") ?? DateTime.Now;
        oBe.Dañado = dr.Field<bool?>("dañado") ?? false;
        oBe.Codigo_bodega_erp = dr.Field<string>("codigo_bodega_erp") ?? "";
        oBe.Sistema = dr.Field<bool?>("sistema") ?? false;
        oBe.Dias_vencimiento_clasificacion = dr.Field<int?>("dias_vencimiento_clasificacion") ?? 0;
        oBe.Tolerancia_dias_vencimiento = dr.Field<int?>("tolerancia_dias_vencimiento") ?? 0;
    }
    public static bool Existe(int IdEstado, SqlConnection conn, SqlTransaction tx)
    {
        const string query = "SELECT COUNT(1) FROM Producto_estado WHERE IdEstado = @IdEstado";
        using var cmd = new SqlCommand(query, conn, tx);
        cmd.Parameters.AddWithValue("@IdEstado", IdEstado);
        return Convert.ToInt32(cmd.ExecuteScalar()) > 0;
    }
    public static int InsertOrUpdate(IConfiguration config, clsBeProducto_estado e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        bool externa = conn != null && tx != null;
        var lConn = externa ? conn! : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTx = null;
        if (!externa) { lConn.Open(); lTx = lConn.BeginTransaction(); }

        try
        {
            if (Existe(e.IdEstado, lConn, externa ? tx! : lTx!))
                return Actualizar(config, e, lConn, externa ? tx : lTx);
            else
                return Insertar(config, e, lConn, externa ? tx : lTx);
        }
        catch (Exception)
        {
            if (!externa && lTx != null) lTx.Rollback();
            throw;
        }
        finally
        {
            if (!externa)
            {
                lTx?.Commit();
                lConn.Close();
            }
        }
    }
    public static int Insertar(IConfiguration config, clsBeProducto_estado e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        Ins.Init("producto_estado");
        Ins.Add("idestado", "@idestado", "F");
        Ins.Add("idpropietario", "@idpropietario", "F");
        Ins.Add("nombre", "@nombre", "F");
        Ins.Add("idubicaciondefecto", "@idubicaciondefecto", "F");
        Ins.Add("utilizable", "@utilizable", "F");
        Ins.Add("activo", "@activo", "F");
        Ins.Add("user_agr", "@user_agr", "F");
        Ins.Add("fec_agr", "@fec_agr", "F");
        Ins.Add("user_mod", "@user_mod", "F");
        Ins.Add("fec_mod", "@fec_mod", "F");
        Ins.Add("dañado", "@dañado", "F");
        Ins.Add("codigo_bodega_erp", "@codigo_bodega_erp", "F");
        Ins.Add("sistema", "@sistema", "F");
        Ins.Add("dias_vencimiento_clasificacion", "@dias_vencimiento_clasificacion", "F");
        Ins.Add("tolerancia_dias_vencimiento", "@tolerancia_dias_vencimiento", "F");

        string sql = Ins.SQL();
        bool externa = conn != null && tx != null;

        using var localConn = externa ? null : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? localTx = null;

        try
        {
            if (!externa)
            {
                localConn!.Open();
                localTx = localConn.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            using var cmd = new SqlCommand(sql, externa ? conn! : localConn!, externa ? tx! : localTx!);
            CrearComando(cmd, e);
            int result = cmd.ExecuteNonQuery();

            if (!externa)
                localTx?.Commit();

            return result;
        }
        catch (Exception ex)
        {
            if (!externa)
                localTx?.Rollback();

            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (!externa && localConn?.State == ConnectionState.Open)
                localConn.Close();
        }
    }
    public static int Actualizar(IConfiguration config, clsBeProducto_estado e, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        Upd.Init("producto_estado");
        Upd.Add("idpropietario", "@idpropietario", "F");
        Upd.Add("nombre", "@nombre", "F");
        Upd.Add("idubicaciondefecto", "@idubicaciondefecto", "F");
        Upd.Add("utilizable", "@utilizable", "F");
        Upd.Add("activo", "@activo", "F");
        Upd.Add("user_agr", "@user_agr", "F");
        Upd.Add("fec_agr", "@fec_agr", "F");
        Upd.Add("user_mod", "@user_mod", "F");
        Upd.Add("fec_mod", "@fec_mod", "F");
        Upd.Add("dañado", "@dañado", "F");
        Upd.Add("codigo_bodega_erp", "@codigo_bodega_erp", "F");
        Upd.Add("sistema", "@sistema", "F");
        Upd.Add("dias_vencimiento_clasificacion", "@dias_vencimiento_clasificacion", "F");
        Upd.Add("tolerancia_dias_vencimiento", "@tolerancia_dias_vencimiento", "F");
        Upd.Where("IdEstado = @IdEstado");

        string sql = Upd.SQL();
        bool externa = conn != null && tx != null;

        using var localConn = externa ? null : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? localTx = null;

        try
        {
            if (!externa)
            {
                localConn!.Open();
                localTx = localConn.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            using var cmd = new SqlCommand(sql, externa ? conn! : localConn!, externa ? tx! : localTx!);
            CrearComando(cmd, e);
            int result = cmd.ExecuteNonQuery();

            if (!externa)
                localTx?.Commit();

            return result;
        }
        catch (Exception ex)
        {
            if (!externa)
                localTx?.Rollback();

            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (!externa && localConn?.State == ConnectionState.Open)
                localConn.Close();
        }
    }
    public static bool GetSingle(IConfiguration config, ref clsBeProducto_estado be)
    {
        const string sql = "SELECT * FROM Producto_estado WHERE IdEstado = @IdEstado";

        using var conn = new SqlConnection(config.GetConnectionString("CST"));
        conn.Open();

        SqlTransaction? tran = null;

        try
        {
            tran = conn.BeginTransaction(IsolationLevel.ReadUncommitted);

            using var cmd = new SqlCommand(sql, conn, tran);
            cmd.Parameters.AddWithValue("@IdEstado", be.IdEstado);

            using var adapter = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            adapter.Fill(dt);

            tran.Commit();

            if (dt.Rows.Count == 1)
            {
                Cargar(ref be, dt.Rows[0]);
                return true;
            }

            return false;
        }
        catch (Exception ex)
        {
            tran?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (conn.State == ConnectionState.Open)
                conn.Close();
        }
    }
    public static List<clsBeProducto_estado> GetAll(IConfiguration config)
    {
        const string sql = "SELECT * FROM Producto_estado";

        using var conn = new SqlConnection(config.GetConnectionString("CST"));
        conn.Open();

        SqlTransaction? tran = null;
        try
        {
            tran = conn.BeginTransaction(IsolationLevel.ReadUncommitted);

            using var cmd = new SqlCommand(sql, conn, tran);
            using var adapter = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            adapter.Fill(dt);

            tran.Commit();

            var list = new List<clsBeProducto_estado>();
            foreach (DataRow row in dt.Rows)
            {
                var item = new clsBeProducto_estado();
                Cargar(ref item, row);
                list.Add(item);
            }
            return list;
        }
        catch (Exception ex)
        {
            tran?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (conn.State == ConnectionState.Open)
                conn.Close();
        }
    }
    public static DataTable Listar(IConfiguration config)
    {
        const string sql = "SELECT * FROM Producto_estado";

        using var conn = new SqlConnection(config.GetConnectionString("CST"));
        conn.Open();

        SqlTransaction? tran = null;
        try
        {
            tran = conn.BeginTransaction(IsolationLevel.ReadUncommitted);
            using var cmd = new SqlCommand(sql, conn, tran);
            using var adapter = new SqlDataAdapter(cmd);
            var dt = new DataTable();
            adapter.Fill(dt);
            tran.Commit();
            return dt;
        }
        catch (Exception ex)
        {
            tran?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (conn.State == ConnectionState.Open)
                conn.Close();
        }
    }
    public static int MaxID(IConfiguration config)
    {
        const string sql = "SELECT ISNULL(MAX(IdEstado), 0) FROM Producto_estado";

        using var conn = new SqlConnection(config.GetConnectionString("CST"));
        conn.Open();

        SqlTransaction? tran = null;
        try
        {
            tran = conn.BeginTransaction(IsolationLevel.ReadUncommitted);
            using var cmd = new SqlCommand(sql, conn, tran);
            object? result = cmd.ExecuteScalar();
            tran.Commit();

            return result != null && result != DBNull.Value ? Convert.ToInt32(result) : 0;
        }
        catch (Exception ex)
        {
            tran?.Rollback();
            var method = new StackTrace().GetFrame(0)?.GetMethod();
            throw new Exception($"{method?.DeclaringType?.Name}.{method?.Name} → {ex.Message}", ex);
        }
        finally
        {
            if (conn.State == ConnectionState.Open)
                conn.Close();
        }
    }
    private static void CrearComando(SqlCommand cmd, clsBeProducto_estado e)
    {
        cmd.Parameters.AddWithValue("@IdEstado", e.IdEstado);
        cmd.Parameters.AddWithValue("@IdPropietario", e.IdPropietario);
        cmd.Parameters.AddWithValue("@nombre", e.Nombre);
        cmd.Parameters.AddWithValue("@IdUbicacionDefecto", e.IdUbicacionDefecto);
        cmd.Parameters.AddWithValue("@utilizable", e.Utilizable);
        cmd.Parameters.AddWithValue("@activo", e.Activo);
        cmd.Parameters.AddWithValue("@user_agr", e.User_agr);
        cmd.Parameters.AddWithValue("@fec_agr", e.Fec_agr);
        cmd.Parameters.AddWithValue("@user_mod", e.User_mod);
        cmd.Parameters.AddWithValue("@fec_mod", e.Fec_mod);
        cmd.Parameters.AddWithValue("@dañado", e.Dañado);
        cmd.Parameters.AddWithValue("@codigo_bodega_erp", e.Codigo_bodega_erp);
        cmd.Parameters.AddWithValue("@sistema", e.Sistema);
        cmd.Parameters.AddWithValue("@dias_vencimiento_clasificacion", e.Dias_vencimiento_clasificacion);
        cmd.Parameters.AddWithValue("@tolerancia_dias_vencimiento", e.Tolerancia_dias_vencimiento);
    }
    public static int MaxID(IConfiguration config, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        const string sql = "SELECT ISNULL(MAX(IdEstado), 0) FROM Producto_estado";
        bool externa = conn != null && tx != null;

        var lConn = externa ? conn! : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTx = null;
        if (!externa) { lConn.Open(); lTx = lConn.BeginTransaction(); }

        try
        {
            using var cmd = new SqlCommand(sql, lConn, externa ? tx! : lTx!);
            var result = cmd.ExecuteScalar();
            if (!externa) lTx?.Commit();
            return Convert.ToInt32(result);
        }
        catch
        {
            if (!externa && lTx != null) lTx.Rollback();
            throw;
        }
        finally
        {
            if (!externa) lConn.Close();
        }
    }
    public static int InsertOrUpdate(IConfiguration config, List<clsBeProducto_estado> estados, SqlConnection? conn = null, SqlTransaction? tx = null)
    {
        bool externa = conn != null && tx != null;
        var lConn = externa ? conn! : new SqlConnection(config.GetConnectionString("CST"));
        SqlTransaction? lTx = null;
        int total = 0;

        try
        {
            if (!externa)
            {
                lConn.Open();
                lTx = lConn.BeginTransaction(IsolationLevel.ReadUncommitted);
            }

            foreach (var e in estados)
            {
                if (Existe(e.IdEstado, lConn, externa ? tx! : lTx!))
                    total += Actualizar(config, e, lConn, externa ? tx : lTx);
                else
                    total += Insertar(config, e, lConn, externa ? tx : lTx);
            }

            if (!externa)
                lTx?.Commit();

            return total;
        }
        catch (Exception)
        {
            if (!externa && lTx != null)
                lTx.Rollback();
            throw;
        }
        finally
        {
            if (!externa)
            {
                lConn.Close();
                lConn.Dispose();
                lTx?.Dispose();
            }
        }
    }
}

File: TOMIMSV4/DAL/Mantenimientos/Producto/ProductoEstado/clsLnProducto_estado.vb
Imports System.Data.SqlClient
Imports System.Reflection

Public Class clsLnProducto_estado

    Public Shared Sub Cargar(ByRef oBeProducto_estado As clsBeProducto_estado, ByRef dr As DataRow, Optional ByVal CargarUbicacionDefectoBodega As Boolean = False)

        Try

            With oBeProducto_estado
                .IdEstado = IIf(IsDBNull(dr.Item("IdEstado")), 0, dr.Item("IdEstado"))
                .IdPropietario = IIf(IsDBNull(dr.Item("IdPropietario")), 0, dr.Item("IdPropietario"))
                .Nombre = IIf(IsDBNull(dr.Item("nombre")), "", dr.Item("nombre"))
                .IdUbicacionDefecto = IIf(IsDBNull(dr.Item("IdUbicacionDefecto")), 0, dr.Item("IdUbicacionDefecto"))
                If CargarUbicacionDefectoBodega Then
                    .IdUbicacionBodegaDefecto = IIf(IsDBNull(dr.Item("IdUbicacionBodegaDefecto")), 0, dr.Item("IdUbicacionBodegaDefecto"))
                End If
                .Utilizable = IIf(IsDBNull(dr.Item("utilizable")), False, dr.Item("utilizable"))
                .Activo = IIf(IsDBNull(dr.Item("activo")), False, dr.Item("activo"))
                .User_agr = IIf(IsDBNull(dr.Item("user_agr")), "", dr.Item("user_agr"))
                .Fec_agr = IIf(IsDBNull(dr.Item("fec_agr")), Date.Now, dr.Item("fec_agr"))
                .User_mod = IIf(IsDBNull(dr.Item("user_mod")), "", dr.Item("user_mod"))
                .Fec_mod = IIf(IsDBNull(dr.Item("fec_mod")), Date.Now, dr.Item("fec_mod"))
                .Dañado = IIf(IsDBNull(dr.Item("Dañado")), False, dr.Item("Dañado"))
                '#CKFK20220704 Agregué que se cargue el campo sistema en la clase
                .Sistema = IIf(IsDBNull(dr.Item("Sistema")), False, dr.Item("Sistema"))
                .Codigo_Bodega_ERP = IIf(IsDBNull(dr.Item("Codigo_Bodega_ERP")), "", dr.Item("Codigo_Bodega_ERP"))
                .Dias_Vencimiento_Clasificacion = IIf(IsDBNull(dr.Item("Dias_Vencimiento_Clasificacion")), 0, dr.Item("Dias_Vencimiento_Clasificacion"))
                .Tolerancia_Dias_Vencimiento = IIf(IsDBNull(dr.Item("Tolerancia_Dias_Vencimiento")), 0, dr.Item("Tolerancia_Dias_Vencimiento"))
                .Reservar_En_UmBas = IIf(IsDBNull(dr.Item("Reserva_En_UmBas")), False, dr.Item("Reserva_En_UmBas"))
            End With

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Sub

    Public Shared Function Insertar(ByRef oBeProducto_estado As clsBeProducto_estado, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Ins.Init("producto_estado")
            Ins.Add("idestado", "@idestado", DataType.Parametro)
            Ins.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Ins.Add("nombre", "@nombre", DataType.Parametro)
            'Ins.Add("idubicaciondefecto","@idubicaciondefecto", DataType.Parametro)
            Ins.Add("utilizable", "@utilizable", DataType.Parametro)
            Ins.Add("activo", "@activo", DataType.Parametro)
            Ins.Add("user_agr", "@user_agr", DataType.Parametro)
            Ins.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Ins.Add("user_mod", "@user_mod", DataType.Parametro)
            Ins.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Ins.Add("dañado", "@dañado", DataType.Parametro)
            Ins.Add("sistema", "@sistema", DataType.Parametro)
            Ins.Add("codigo_bodega_erp", "@codigo_bodega_erp", DataType.Parametro)
            Ins.Add("dias_vencimiento_clasificacion", "@dias_vencimiento_clasificacion", DataType.Parametro)
            Ins.Add("tolerancia_dias_vencimiento", "@tolerancia_dias_vencimiento", DataType.Parametro)
            Ins.Add("reserva_en_umbas", "@reserva_en_umbas", DataType.Parametro)

            Dim sp As String = Ins.SQL()
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDESTADO", oBeProducto_estado.IdEstado))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeProducto_estado.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", clsPublic.Quitar_Caracteres_No_Permitidos(oBeProducto_estado.Nombre)))
            cmd.Parameters.Add(New SqlParameter("@UTILIZABLE", oBeProducto_estado.Utilizable))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_estado.Activo))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_estado.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_estado.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_estado.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_estado.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@DAÑADO", oBeProducto_estado.Dañado))
            cmd.Parameters.Add(New SqlParameter("@SISTEMA", oBeProducto_estado.Sistema))
            cmd.Parameters.Add(New SqlParameter("@CODIGO_BODEGA_ERP", oBeProducto_estado.Codigo_Bodega_ERP))
            cmd.Parameters.Add(New SqlParameter("@DIAS_VENCIMIENTO_CLASIFICACION", oBeProducto_estado.Dias_Vencimiento_Clasificacion))
            cmd.Parameters.Add(New SqlParameter("@TOLERANCIA_DIAS_VENCIMIENTO", oBeProducto_estado.Tolerancia_Dias_Vencimiento))
            cmd.Parameters.Add(New SqlParameter("@RESERVA_EN_UMBAS", oBeProducto_estado.Reservar_En_UmBas))
            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
        End Try

    End Function

    Public Shared Function Actualizar(ByRef oBeProducto_estado As clsBeProducto_estado, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Upd.Init("producto_estado")
            Upd.Add("idestado", "@idestado", DataType.Parametro)
            Upd.Add("idpropietario", "@idpropietario", DataType.Parametro)
            Upd.Add("nombre", "@nombre", DataType.Parametro)
            Upd.Add("idubicaciondefecto", "@idubicaciondefecto", DataType.Parametro)
            Upd.Add("utilizable", "@utilizable", DataType.Parametro)
            Upd.Add("activo", "@activo", DataType.Parametro)
            Upd.Add("user_agr", "@user_agr", DataType.Parametro)
            Upd.Add("fec_agr", "@fec_agr", DataType.Parametro)
            Upd.Add("user_mod", "@user_mod", DataType.Parametro)
            Upd.Add("fec_mod", "@fec_mod", DataType.Parametro)
            Upd.Add("Dañado", "@Dañado", DataType.Parametro)
            Upd.Add("sistema", "@sistema", DataType.Parametro)
            Upd.Add("codigo_bodega_erp", "@codigo_bodega_erp", DataType.Parametro)
            Upd.Add("dias_vencimiento_clasificacion", "@dias_vencimiento_clasificacion", DataType.Parametro)
            Upd.Add("tolerancia_dias_vencimiento", "@tolerancia_dias_vencimiento", DataType.Parametro)
            Upd.Add("reserva_en_umbas", "@reserva_en_umbas", DataType.Parametro)
            Upd.Where("IdEstado = @IdEstado")

            Dim sp As String = Upd.SQL()

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDESTADO", oBeProducto_estado.IdEstado))
            cmd.Parameters.Add(New SqlParameter("@IDPROPIETARIO", oBeProducto_estado.IdPropietario))
            cmd.Parameters.Add(New SqlParameter("@NOMBRE", clsPublic.Quitar_Caracteres_No_Permitidos(oBeProducto_estado.Nombre)))
            cmd.Parameters.Add(New SqlParameter("@IDUBICACIONDEFECTO", oBeProducto_estado.IdUbicacionDefecto))
            cmd.Parameters.Add(New SqlParameter("@UTILIZABLE", oBeProducto_estado.Utilizable))
            cmd.Parameters.Add(New SqlParameter("@ACTIVO", oBeProducto_estado.Activo))
            cmd.Parameters.Add(New SqlParameter("@USER_AGR", oBeProducto_estado.User_agr))
            cmd.Parameters.Add(New SqlParameter("@FEC_AGR", oBeProducto_estado.Fec_agr))
            cmd.Parameters.Add(New SqlParameter("@USER_MOD", oBeProducto_estado.User_mod))
            cmd.Parameters.Add(New SqlParameter("@FEC_MOD", oBeProducto_estado.Fec_mod))
            cmd.Parameters.Add(New SqlParameter("@DAÑADO", oBeProducto_estado.Dañado))
            cmd.Parameters.Add(New SqlParameter("@SISTEMA", oBeProducto_estado.Sistema))
            cmd.Parameters.Add(New SqlParameter("@CODIGO_BODEGA_ERP", oBeProducto_estado.Codigo_Bodega_ERP))
            cmd.Parameters.Add(New SqlParameter("@DIAS_VENCIMIENTO_CLASIFICACION", oBeProducto_estado.Dias_Vencimiento_Clasificacion))
            cmd.Parameters.Add(New SqlParameter("@TOLERANCIA_DIAS_VENCIMIENTO", oBeProducto_estado.Tolerancia_Dias_Vencimiento))
            cmd.Parameters.Add(New SqlParameter("@RESERVA_EN_UMBAS", oBeProducto_estado.Reservar_En_UmBas))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex1
        Catch ex As Exception
            If lTransaction IsNot Nothing Then lTransaction.Rollback()
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
        End Try

    End Function

    Public Shared Function Eliminar(ByRef oBeProducto_estado As clsBeProducto_estado, Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Producto_estado" &
             "  Where(IdEstado = @IdEstado)"

            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection, pTransaction)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            cmd.Parameters.Add(New SqlParameter("@IDESTADO", oBeProducto_estado.IdEstado))

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
        End Try
    End Function

    Public Shared Function EliminarTodos(Optional ByVal pConection As SqlConnection = Nothing, Optional ByVal pTransaction As SqlTransaction = Nothing) As Integer

        Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
        Dim lTransaction As SqlTransaction = Nothing

        Try

            Const sp As String = " Delete from Producto_estado"
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim Es_Transaccion_Remota As Boolean = (pConection IsNot Nothing AndAlso pTransaction IsNot Nothing)

            If Es_Transaccion_Remota Then
                cmd = New SqlCommand(sp, pConection)
            Else
                lConnection.Open() : lTransaction = lConnection.BeginTransaction(IsolationLevel.ReadCommitted)
                cmd = New SqlCommand(sp, lConnection, lTransaction)
            End If

            Dim rowsAffected As Integer = cmd.ExecuteNonQuery()

            cmd.Dispose()

            If Not Es_Transaccion_Remota Then lTransaction.Commit()

            Return rowsAffected

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Throw ex
        Finally
            If lConnection.State = ConnectionState.Open Then lConnection.Close()
        End Try

    End Function

    Public Shared Function Listar() As DataTable

        Try

            Const sp As String = "SELECT * FROM Producto_estado"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable
            dad.Fill(dt)

            Return dt

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function Obtener(ByRef oBeProducto_estado As clsBeProducto_estado) As Boolean

        Try

            Const sp As String = "SELECT * FROM Producto_estado" &
            " Where(IdEstado = @IdEstado)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)

            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDESTADO", oBeProducto_estado.IdEstado))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(oBeProducto_estado, dt.Rows(0))
            Else
                Throw New Exception("No se pudo obtener el registro")
            End If

            Return True

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function GetAll() As List(Of clsBeProducto_estado)

        Try

            Dim lReturnList As New List(Of clsBeProducto_estado)
            Const sp As String = "SELECT * FROM Producto_estado"
            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            Dim dt As New DataTable

            dad.Fill(dt)

            Dim vBeProducto_estado As New clsBeProducto_estado

            For Each dr As DataRow In dt.Rows

                vBeProducto_estado = New clsBeProducto_estado
                Cargar(vBeProducto_estado, dr)
                lReturnList.Add(vBeProducto_estado)

            Next

            lConnection.Dispose()
            cmd.Dispose()

            Return lReturnList

        Catch ex As Exception
            Throw ex
        End Try

    End Function

    Public Shared Function GetSingle(ByRef pBeProducto_estado As clsBeProducto_estado)

        Try

            Const sp As String = "SELECT * FROM Producto_estado" &
            " Where(IdEstado = @IdEstado)"

            Dim lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
            Dim cmd As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
            Dim dad As New SqlDataAdapter(cmd)
            dad.SelectCommand.Parameters.Add(New SqlParameter("@IDESTADO", pBeProducto_estado.IdEstado))

            Dim dt As New DataTable
            dad.Fill(dt)

            If dt.Rows.Count = 1 Then
                Cargar(pBeProducto_estado, dt.Rows(0))
            End If

            Return True

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

    Public Shared Function MaxID() As Integer

        Try

            Dim lMax As Integer = 0

            Const sp As String = "SELECT ISNULL(Max(IdEstado),0) FROM Producto_estado"

            Using lConnection As New SqlConnection(Configuration.ConfigurationManager.AppSettings("CST"))
                lConnection.Open()
                Using lCommand As New SqlCommand(sp, lConnection) With {.CommandType = CommandType.Text}
                    Dim lReturnValue As Object = lCommand.ExecuteScalar()
                    lConnection.Close()
                    If lReturnValue IsNot DBNull.Value AndAlso lReturnValue IsNot Nothing Then
                        lMax = CInt(lReturnValue)
                    End If
                End Using
            End Using

            Return lMax

        Catch ex1 As SqlException
            Throw ex1
        Catch ex As Exception
            Dim vMsgError As String = String.Format("{0} {1}", MethodBase.GetCurrentMethod.Name(), ex.Message)
            clsLnLog_error_wms.Agregar_Error(vMsgError)
            Throw ex
        End Try

    End Function

End Class